### DISCUSSION TITLE
Improve user experience by adjusting --limit-results behavior

### DISCUSSION THREAD

**Discussion by kirillbobyrev** [clangd/clangd#761] [Ideas]
Based on several discussions and user questions in Discord, the limit on number of items returned by queries (controlled by `--limit-results` and set to `100` by default) is really unexpected and confusing to the users. What typically happens is that people try find-refs and then are very surprised that it doesn't find all references they can find through sheer `grep`/`rg` plain text search. Some think Clangd is not working correctly/has bugs and hence doesn't really know what the references are. Here are some ideas I thought could improve UX and make the default behavior more useful & convenient for our users.

## Transparency

Right now the `--limit-results` flag documentation says

> --limit-results=<int>           - Limit the number of results returned by clangd. 0 means no limit (default=100)

I actually didn't know if we had the flag and had to `rg 100` inside `clang-tools-extra/clangd` because I remember we had a limit of 100 back in the Dex days but I didn't know how exactly it is controlled :) If it wasn't obvious to me, then it's unlikely it'd be obvious for most users. And when I found the flag, I was a little confused by the documentation. I know the "results" mean CC items, refs, etc but I had to look through the code real quick to confirm this, so I guess most users don't.

Also, the users don't find it obvious that we have any kind of limit at all because I think the concept of improving performance by reducing the number of returned results is only apparent from the implementation perspective :)

Potential actionable items here:

- Adding some kind of visible documentation explaining this concept? **Actually, should we have an FAQ?** Some questions seem to be repeating themselves, so maybe it makes sense to provide the answers to those and point people to the FAQ so that they're comfortable with it/know about its existance?
- Possibly change the wording/add explanation to the `--limit-results` flag?
- Add some kind of warning in the logs, saying something like "hey, you might want to get all the references but 

## Refs request

[As of right now](https://github.com/llvm/llvm-project/blob/60ed86d3507bf908e0079f78b246ae096321ae03/clang-tools-extra/clangd/ClangdLSPServer.cpp#L1265), we do the following for "find refs" request:

```c++
  Server->findReferences(
      Params.textDocument.uri.file(), Params.position, Opts.CodeComplete.Limit,
```

using `Opts.CodeComplete.Limit` for limiting the number of returned references. This is non-intuitive even from the implementation perspective.

Potential actionable items:

- Split the `--limit-results` flag into `--limit-cc-results` and `--limit-refs-reults` (names are non-optimal but you get the point). The idea is to give users more control over limiting the number of returned references vs code-completion performance. I can imagine scenarios when one wants the CC to be fast (because it affects overall latency/responsiveness) but doesn't mind spending some time to get the refs.
- Code Search is only available for few projects (such as Chromium) so it's natural for people to desire something like that for their project. Clangd does CS's "job" in most cases, and one of the most important cases is actually the "find references" request. I think the references request is the one where the vast majority of our users doesn't mind to spend some time and wait N seconds for the sake of getting the complete list of the results. We can have a timeout/warning if it takes too long and restart the request with the limit.

## Remote index server

`clangd-index-server` is supposed to be running on powerful machines. Therefore, should we just ignore the limit at least for the refs request and/or increase it from the default `100` for code completion etc? I know the latter will affect the client performance, too, but I don't know how expensive is the client-side ranking (the decision forest logic now, right?).

Potential actionable items:

- Add limit controlling flags for `clangd-index-server` binary. Potential flags:
  - `--override-refs-limit=`
  - `--override-cc-limit=`
- Remove the references limit altogether? We can observe the request latency in this case and see if it is okay. The remote index server is supposed to provide index for really large projects (otherwise it wouldn't be needed) but I think it also makes the expectations for refs to be Code Search-like.

## Dynamic limit adjustment

This one probably requires thorough investigation and some well-thought solution but I just stumbled across the idea: I think we have enough room for performance in the index request space (dex is still holding up!) and we might want to experiment here with adjusting the results limit dynamically.

### Simple heuristics

Here's a basic idea: if the project doesn't have more than X files/(or even better) if the index is not larger than X MB, then we can increase the limit. The code completion quality would potentially increase (though probably not by much) and the users won't even notice. This is quite simple and would make a good default, would not require users to have much thought and probably have a positive effect right away. Also, really simple to implement.

### Complicated heuristics

There is also a room for some complicated ideas: observing the average request completion time and trying to keep it below some specific number of ms. We can try adjusting the limit at runtime by continuously measure that latency and tweak the limit so that it satisfies our performance requirements. Those requirements can also probably be tweaked, e.g. if the user sets `--limit-results-strategy=aggressive` that would mean they want to keep the latency below 2 ms on average, if `--limit-results-strategy=maximum-quality` - then they'd be fine with having a slower editor but better quality.

This probably requires a lot of measurements and careful design. The implementation is also unlikely to be very straightforward. Another drawback is that it removes the transparency from the process and the users might be unable to follow what's happening without a deep understanding of internals/good documentation with enough visibility and context.

**Comment by HighCommander4** [comment]
+1 for treating code completion and find-references differently here. With code completion, the user's going to give up and type out the name long before they scroll down to the 100th result, but for find-references, you often do want to see all of them, e.g. if you're refactoring a function or just auditing to see how its callers use it.

In my mind, the natural UX for find-refs would be to provide the results incrementally (with no limit on the total). It looks like LSP has support for this in the form of [partial results](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#partialResults)?

**Reply by kirillbobyrev** [reply]
Oh, partial results are a cool concept. Yes, what I wanted to write down along the lines is that it'd also be nice to somehow indicate the "progress" (loading bar or whatever) to indicate that we're actually doing something so that the user doesn't give up. The partial results is probably even better!

**Reply by sam-mccall** [reply]
+1 to increasing the refs limit.

I don't think it should be infinity, it's easy to (mistakenly?) find-refs on `std::string` in a huge codebase. I can't see a scenario where serving say 100K results to the client would be helpful, but I can easily imagine cases where it breaks something.

Regarding streaming results, I think we should be a bit cautious:
 - it requires a bunch of plumbing/framework work
 - it won't be supported by all (or I suspect most) clients, so we *still* need good answers for the non-streaming case too
 - there's not very much index-side work to be done between first & last result in typical cases, so all results will get shoveled into the channel before the client has a realistic chance to cancel. (This is mostly true for local indexes, less so for remote as GRPC provides pushback)
 - the SymbolIndex interface is streaming, but composing them (MergeIndex) requires us to buffer entirety of results from N-1 indexes. Research needed to know if this can be addressed. If we must backload all the results, there's little point.

If we can pick a limit K so that we can serve K results with good latency and results >K aren't interesting, then that seems preferable to me. 1000 or 10000 seem like reasonable candidates, maybe I lack imagination.

(It would be nice to indicate somehow if refs are incomplete. showMessage maybe? But note that the index can be incomplete, and we don't have a good way to detect that)

**Reply by kirillbobyrev** [reply]
> I don't think it should be infinity, it's easy to (mistakenly?) find-refs on std::string in a huge codebase. I can't see a scenario where serving say 100K results to the client would be helpful, but I can easily imagine cases where it breaks something.

That's been my concern too, but I've tried setting `--limit-results=0` and living with that for a couple days over the weekend and it seems reasonable to me. E.g. getting 25,385 `std::string` references in LLVM takes (almost) no time and the UI doesn't hang (surprisingly), so I don't think this is as scary as I imagined it to be. Of course, I can imagine this blowing up somehow but there are arguments on "well, I guess one shouldn't request `std::string` references in projects with 900000000M LOC". But then also

> We can have a timeout/warning if it takes too long and restart the request with the limit.

Issuing a warning "this takes too long, restarting with 10k limit" doesn't sound awful to me.

**Reply by sam-mccall** [reply]
> That's been my concern too, but I've tried setting --limit-results=0 and living with that for a couple days over the weekend and it seems reasonable to me

Sure, it works on your machine :-)
Have you tried running over a tethered mobile connection? Using a client with a slow JSON parser or bad location list, or one that fetches snippets from files on disk synchronously? Working on a project 10-20x the size of LLVM?

Infinity is an aesthetically pleasing limit, but is there any **other** practical reason to make it the default :-)

> Issuing a warning "this takes too long, restarting with 10k limit" doesn't sound awful to me.

This wouldn't mitigate most of the issues (it's more likely to be the client/transport that's slow than the server). And it's complicated to implement.

I think there are lots of interesting ideas in this thread, but if we're rejecting a simple solution in favor of a more sophisticated one, we need to know why the simple solution is bad, not just how the complicated one can be made to work.

**Reply by kirillbobyrev** [reply]
> Sure, it works on your machine :-)
> Have you tried running over a tethered mobile connection? Using a client with a slow JSON parser or bad location list, or one that fetches snippets from files on disk synchronously? Working on a project 10-20x the size of LLVM?

I've also tried on the MacBook that is a bit sloppy (takes over 10 mins just to fetch LLVM sources and many hours just to build Clangd) and my personal machine that might be even more sloppy :D Both work great. Even when I have over 10k refs the UI really seems flawless and picture-perfect. I really believe we shouldn't limit refs, especially for the remote index.

I think the argument here is not only that it's aesthetically pleasing but that I do not see any regressions testing on different devices and under different circumstances. Sure, it _might_ blow up on some underpowered machines with the developers working on a project 9000 times bigger than LLVM but I'm not sure we should be optimizing for them (and for those machines we would already have other issues that are arguably much more important).

**Reply by kadircet** [reply]
For another sample point, i've tried hitting xrefs for std::string in chromium codebase. It yielded ~128k results in 1 sec (due to timeout on the streaming RPC). Afterwards it took about 3 secs for clangd to serialize those results to stdout. Then my LSP plugin (YCM) took about 20 secs to deserialize clangd's output and populate location list. I was using a pretty decent workstation (with a high bandwidth internet connection), though all of this is handled by a single thread and mostly IO bound, so I am not sure if the power of the computer matters a lot.

On the server side query resulted in ~350k results and took 1.8 secs to resolve (excluding network latency) `[public] request v1/Refs => OK: 349179 results in 1766ms`. To transfer the reply we've used ~8MB of network traffic (our daily average for outgoing traffic is ~60MB). As for CPU we've utilized about 3 times more resources than our daily average. (There aren't any implications on disk io, as we already load everything into memory).

So my 2 cents on flawless UI, it is probably because you have a smarter editor/lsp plugin that can incrementally load results into UI rather than filling them all at once. Also, network IO is currently limited to 1 second. So people will get different results depending on their connection speed, and if you were testing with a slow connection you would end up fetching less results making the experience feel flawless (but incomplete).

Hence I am also hesitant on making the limit infinite, but other ideas in the doc looks really promising. Another dimension to look into could also be about coverage in terms of files. For example we can return full set of refs for active files (or just main file) and then try to display results from as many different files as possible (bonus points if we can tell something like +1234 results in this file). That would've at least allow me to have more trust on the tool, but it is just my opinion.

**Comment by kirillbobyrev** [comment]
Also, there is the relations request but I'm not really sure about its use cases and how this all applies to that one. Thoughts?

**Reply by sam-mccall** [reply]
Relations is used to find subclasses, overrides etc.
It's not likely we're hitting the limit there regularly.
